libtenzir/builtins/operators/buffer.cpp

```diff
@@ -0,0 +1,408 @@
+          });
+      co_yield {};
+    }
+  }
+
+  auto name() const -> std::string override {
+    return "internal-write-buffer";
+  }
+
+  auto optimize(expression const& filter, event_order order) const
+    -> optimize_result override {
+    return optimize_result{filter, order, copy()};
+  }
+
+  auto infer_type_impl(operator_type input) const
+    -> caf::expected<operator_type> override {
+    if (input.is<table_slice>()) {
+      return tag_v<table_slice>;
+    }
+    return diagnostic::error("`buffer` does not accept {} as input",
+                             operator_type_name(input))
+      .to_error();
+  }
+
+  friend auto inspect(auto& f, write_buffer_operator& x) -> bool {
+    return f.object(x).fields(f.field("id", x.id_));
+  }
+
+private:
+  uuid id_ = {};
+};
+
+class read_buffer_operator final : public crtp_operator<read_buffer_operator> {
+public:
+  read_buffer_operator() = default;
+
+  explicit read_buffer_operator(uuid id, located<uint64_t> capacity,
+                                std::optional<buffer_policy> policy)
+    : id_{id}, capacity_{capacity}, policy_{policy} {
+  }
+
+  auto policy(operator_control_plane& ctrl) const -> buffer_policy {
+    return policy_.value_or(ctrl.is_hidden() ? buffer_policy::drop
+                                             : buffer_policy::block);
+  }
+
+  auto metrics(operator_control_plane& ctrl) const -> metric_handler {
+    return ctrl.metrics(type{
+      "tenzir.metrics.buffer",
+      record_type{
+        {"used", uint64_type{}},
+        {"remaining", uint64_type{}},
+        {"dropped", uint64_type{}},
+      },
+    });
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    // The internal-read-buffer operator is spawned before the
+    // internal-write-buffer operator, so we spawn the buffer actor here and
+    // move it into the registry before the first yield.
+    auto buffer = ctrl.self().spawn<caf::linked>(make_buffer, capacity_,
+                                                 policy(ctrl), metrics(ctrl),
+                                                 ctrl.shared_diagnostics());
+    ctrl.self().system().registry().put(fmt::format("tenzir.buffer.{}", id_),
+                                        buffer);
+    co_yield {};
+    // Now, we can get batch by batch from the buffer.
+    for (auto&& events : input) {
+      TENZIR_ASSERT(events.rows() == 0);
+      ctrl.set_waiting(true);
+      ctrl.self()
+        .request(buffer, caf::infinite, atom::read_v)
+        .then(
+          [&](table_slice& response) {
+            ctrl.set_waiting(false);
+            events = std::move(response);
+          },
+          [&](caf::error& err) {
+            diagnostic::error(err)
+              .note("failed to read from buffer")
+              .emit(ctrl.diagnostics());
+          });
+      co_yield {};
+      if (events.rows() == 0) {
+        co_return;
+      }
```

Comment:
```suggestion
      if (events.rows() == 0) {
        continue;
      }
```

Small change, big effect.

Reply:
That's the shutdown path, so this was intended.

Reply:
Doesn't seem to work correctly though. With the `co_return`, my `source {} | repeat | buffer 1, policy="drop" | head 10` pipeline does not return 10 events.

Reply:
Hm, this may actually be caused by me doing the shutdown via linking. Might have to find a different approach then that's event-driven.

Reply:
Figured it out, we sometimes split at 0 when the buffer was exactly at capacity, leading to an early shutdown. I'll fix that and simplify this.

Reply:
Fixed in 32a37c213d1; turned out to be even more simple this way.